src/intrinsics/ecma262/Object.js

```diff
@@ -101,36 +136,69 @@ export default function(realm: Realm): NativeFunctionValue {
         }
         keys = frm.$OwnPropertyKeys();
 
-        if (to_must_be_partial) {
-          if (to instanceof AbstractObjectValue && to.values.isTop()) {
-            // We don't know which objects to make partial and making all objects partial is failure in itself
-            AbstractValue.reportIntrospectionError(to);
-            throw new FatalError();
-          } else {
-            // if to has properties, we better remove them because after the temporal call to Object.assign we don't know their values anymore
-            if (to.hasStringOrSymbolProperties()) {
-              // preserve them in a snapshot and add the snapshot to the sources
-              delayedSources.push(to.getSnapshot({ removeProperties: true }));
-            }
+        handleSnapshot(frm, frm_was_partial);
+
+        // c. Repeat for each element nextKey of keys in List order,
+        invariant(frm, "from required");
+        invariant(keys, "keys required");
+        copyKeys(keys, frm, to);
+      };
+
+      const tryAndApplySourceOrRecover = nextSource => {
+        let effects;
+        let savedSuppressDiagnostics = realm.suppressDiagnostics;
+        try {
+          realm.suppressDiagnostics = true;
+          effects = realm.evaluateForEffects(
+            () => {
+              applySource(nextSource);
+              return realm.intrinsics.undefined;
+            },
+            undefined,
+            "tryAndApplySourceOrRecover"
+          );
+        } catch (e) {
+          // ignore null or undefined
+          if (nextSource === realm.intrinsics.null || nextSource === realm.intrinsics.undefined) {
+            return;
+          }
+          let frm = To.ToObject(realm, nextSource);
+
+          if (e instanceof FatalError && to.isSimpleObject()) {
+            let frm_was_partial = frm.isPartialObject();
 
             if (frm_was_partial) {
-              if (frm instanceof AbstractObjectValue && frm.kind === "explicit conversion to object") {
-                // Make it implicit again since it is getting delayed into an Object.assign call.
-                delayedSources.push(frm.args[0]);
-              } else {
-                frmSnapshot = frm.getSnapshot();
-                frm.temporalAlias = frmSnapshot;
-                frm.makePartial();
-                delayedSources.push(frmSnapshot);
-              }
+              to_must_be_partial = true;
             }
+            handleSnapshot(frm, frm_was_partial);
```

Comment:
So there's a case here that's pretty dangerous.

We can end up in the error handling scenario for any number of different reasons. There are many things that can cause fatals.

Mostly that's fine because we'll just add to the delayedSources. However, not all branches in `handleSnapshot` actually does that. Particularly if this is a not a partial object but throws for some other reason (e.g. some unknown getter gets invoked on a known object) then we end up in recovery mode, but `handleSnapshot` doesn't add it to delayed sources in that case so it ends up just being ignored instead.

I think this is a subtle bug. Probably need to refactor handleSnapshot a bit to ensure that all branches end up adding delayed sources for the recovery case. Which means that it always ends up with a partial to object if we have to recover.